packages/jsii/lib/runtime-info.ts

```diff
@@ -0,0 +1,126 @@
+import * as ts from 'typescript';
+
+/**
+ * Provides a TransformerFactory to annotate classes with runtime information
+ * (e.g., fully-qualified name, version).
+ *
+ * It does this by first inserting this definition at the top of each source file:
+ * ```
+ * var vfqnSym = Symbol.for("jsii.rtti");
+ * ```
+ *
+ * Then, for each class that has registered runtime information during assembly,
+ * insert a static member to the class with its fqn and version:
+ * ```
+ * private static [vfqnSym] = { fqn: "ModuleName.ClassName", version: "1.2.3" }
+ * ```
+ */
+export class RuntimeTypeInfoInjector {
+  private readonly fqnsByClass = new Map<ts.ClassDeclaration, string>();
+
+  public constructor(private readonly version: string) {}
+
+  /**
+   * Register the fully-qualified name (fqn) of a class with its ClassDeclaration.
+   * Only ClassDeclarations with registered fqns will be annotated.
+   */
+  public registerClassFqn(clazz: ts.ClassDeclaration, fqn: string) {
+    this.fqnsByClass.set(clazz, fqn);
+  }
+
+  /**
+   * Return the set of Transformers to be used in TSC's program.emit()
+   */
+  public makeTransformers(): ts.CustomTransformers {
+    return {
+      before: [this.runtimeTypeTransformer()],
+    };
+  }
+
+  public runtimeTypeTransformer(): ts.TransformerFactory<ts.SourceFile> {
+    // eslint-disable-next-line @typescript-eslint/no-this-alias
+    const self = this;
+
+    return (context) => {
+      return (sourceFile) => {
+        const rttiSymbolIdentifier = ts.createIdentifier('vfqnSym');
+        const rttiSymbol = ts.createCall(
+          ts.createPropertyAccess(
+            ts.createIdentifier('Symbol'),
+            ts.createIdentifier('for'),
+          ),
+          undefined,
+          [ts.createStringLiteral('jsii.rtti')],
+        );
+        const rttiSymbolDeclaration = ts.createVariableDeclaration(
+          rttiSymbolIdentifier,
+          undefined,
+          rttiSymbol,
+        );
+
+        sourceFile = ts.updateSourceFileNode(sourceFile, [
+          ts.createVariableStatement([], [rttiSymbolDeclaration]),
+          ...sourceFile.statements,
+        ]);
```

Comment:
Is there a chance this will cause "unused identifier" errors in any source file that does not declare an exported type?

Reply:
Possibly. I didn't run into it in testing, but it's a reasonable call-out. I can adjust to only emit the symbol when necessary.